bus/pruning.go

```diff
@@ -0,0 +1,185 @@
+				Size:             rev.Filesize,
+
+				MissedHostPayout:  rev.MissedHostPayout(),
+				ValidRenterPayout: rev.ValidRenterPayout(),
+			},
+		})
+	}
+
+	var pruneErr string
+	if err != nil {
+		pruneErr = err.Error()
+	}
+
+	return api.ContractPruneResponse{
+		ContractSize: rev.Filesize,
+		Pruned:       pruned,
+		Remaining:    remaining,
+		Error:        pruneErr,
+	}, nil
+}
+
+func (b *Bus) pruneContractV2(ctx context.Context, rk types.PrivateKey, cm api.ContractMetadata, hostIP string, gc gouging.Checker, pendingUploads map[types.Hash256]struct{}) (api.ContractPruneResponse, error) {
+	signer := ibus.NewFormContractSigner(b.w, rk)
+
+	// get latest revision
+	rev, err := b.rhp4Client.LatestRevision(ctx, cm.HostKey, hostIP, cm.ID)
+	if err != nil {
+		return api.ContractPruneResponse{}, fmt.Errorf("failed to fetch revision for pruning: %w", err)
+	} else if rev.RevisionNumber < cm.RevisionNumber {
+		return api.ContractPruneResponse{}, fmt.Errorf("latest known revision %d is less than contract revision %d", rev.RevisionNumber, cm.RevisionNumber)
+	}
+
+	// get prices
+	settings, err := b.rhp4Client.Settings(ctx, cm.HostKey, hostIP)
+	if err != nil {
+		return api.ContractPruneResponse{}, fmt.Errorf("failed to fetch prices for pruning: %w", err)
+	}
+	prices := settings.Prices
+
+	// make sure they are sane
+	if gb := gc.CheckV2(settings); gb.Gouging() {
+		return api.ContractPruneResponse{}, fmt.Errorf("host for pruning is gouging: %v", gb.String())
+	}
+
+	// fetch all contract roots
+	numsectors := rev.Filesize / rhpv4.SectorSize
+	sectorRoots := make([]types.Hash256, 0, numsectors)
+	var rootsUsage rhpv4.Usage
+	for offset := uint64(0); offset < numsectors; {
+		// calculate the batch size
+		length := uint64(rhpv4.MaxSectorBatchSize)
+		if offset+length > numsectors {
+			length = numsectors - offset
+		}
+
+		// fetch the batch
+		res, err := b.rhp4Client.SectorRoots(ctx, cm.HostKey, hostIP, b.cm.TipState(), prices, signer, cRHP4.ContractRevision{
+			ID:       cm.ID,
+			Revision: rev,
+		}, offset, length)
+		if err != nil {
+			return api.ContractPruneResponse{}, err
+		}
+
+		// update revision since it was revised
+		rev = res.Revision
+
+		// collect roots
+		sectorRoots = append(sectorRoots, res.Roots...)
+		offset += uint64(len(res.Roots))
+
+		// update the cost
+		rootsUsage = rootsUsage.Add(res.Usage)
+	}
+
+	// fetch indices to prune
+	indices, err := b.store.PrunableContractRoots(ctx, cm.ID, sectorRoots)
+	if err != nil {
+		return api.ContractPruneResponse{}, err
+	}
+
+	// avoid pruning pending uploads
+	toPrune := indices[:0]
+	for _, index := range indices {
+		_, ok := pendingUploads[sectorRoots[index]]
+		if !ok {
+			toPrune = append(toPrune, index)
+		}
+	}
+	totalToPrune := uint64(len(toPrune))
+
+	// cap at max batch size
+	batchSize := rhpv4.MaxSectorBatchSize
+	if batchSize > len(toPrune) {
+		batchSize = len(toPrune)
+	}
+	toPrune = toPrune[:batchSize]
+
+	// prune the batch
```

Comment:
why don't we free sectors in a loop here?

Reply:
It just makes things a bit simpler/cleaner to make 1 http call == pruning up to 1 batch of sectors. Rather than having a timeout we need to handle, a bunch of additional index adjustment code etc.